packages/legend-query-builder/src/components/fetch-structure/QueryBuilderResultModifierPanel.tsx

```diff
@@ -380,6 +385,283 @@ export const QueryResultModifierModal = observer(
+            exculdeMilestoningState: true,
+          })
+        ) {
+          resetParameter(undefined);
+          return (lambdaParamState: LambdaParameterState) =>
+            lambdaParamState.parameter.name !== param.name;
+        }
+      }
+      return (lambdaParamState: LambdaParameterState) => true;
+    };
+
+    const setAllVersions = (value: boolean | undefined): void => {
+      if (value) {
+        //clean all unused milestoning parameters e.g. businessDate, ..., endDate.
+        setParameterStates([
+          ...parameterStates.filter(
+            (ps) =>
+              shouldFilterMilestoningParamIfNotUsed(
+                businessDate,
+                setBusinessDate,
+              )(ps) &&
+              shouldFilterMilestoningParamIfNotUsed(
+                processingDate,
+                setProcessingDate,
+              )(ps) &&
+              shouldFilterMilestoningParamIfNotUsed(
+                startDate,
+                setStartDate,
+              )(ps) &&
+              shouldFilterMilestoningParamIfNotUsed(endDate, setEndDate)(ps),
+          ),
+        ]);
+      } else {
+        //get or initialize getAll() parameters
+        setIsAllVersionsInRangeEnabled(false);
+        let newParamStates: LambdaParameterState[] = [];
+        if (
+          (!businessDate || !processingDate) &&
+          milestoningState.queryBuilderState.class
+        ) {
+          const stereotype = getMilestoneTemporalStereotype(
+            milestoningState.queryBuilderState.class,
+            milestoningState.queryBuilderState.graphManagerState.graph,
+          );
+          if (stereotype) {
+            const existingParamStates = parameterStates.map(
+              (ps) => ps.variableName,
+            );
+            newParamStates = milestoningState
+              .getMilestoningImplementation(stereotype)
+              .buildParameterStatesFromMilestoningParameters()
+              .filter((ps) => !existingParamStates.includes(ps.parameter.name));
+          }
+        }
+        const allParamStates = [...newParamStates, ...parameterStates].filter(
+          (ps) =>
+            shouldFilterMilestoningParamIfNotUsed(
+              startDate,
+              setStartDate,
+            )(ps) &&
+            shouldFilterMilestoningParamIfNotUsed(endDate, setEndDate)(ps),
+        );
+        setParameterStates(allParamStates);
+        if (!businessDate) {
+          setBusinessDate(
+            allParamStates.find(
+              (ps) =>
+                ps.variableName ===
+                (milestoningState.businessDate &&
+                milestoningState.businessDate instanceof VariableExpression
+                  ? milestoningState.businessDate.name
+                  : BUSINESS_DATE_MILESTONING_PROPERTY_NAME),
+            )?.parameter,
+          );
+        }
+        if (!processingDate) {
+          setProcessingDate(
+            allParamStates.find(
```

Comment:
Can you check if it works for the case when the root class is not milestoned but one of the property in projection is milestoned. I see that you are updating newparamstates based on root class milestoning type, then it wouldn't have any paramter named processingDate or businessdate to set these values if one of the property in projection is milestoned.

Reply:
yeah this case also works as I keep any used milestoning parameters 

https://github.com/finos/legend-studio/assets/73408381/49a96122-b0af-4899-8c0e-b4a81d42de9b